src/aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -192,74 +198,134 @@ def setup(self):
+           are still Pulay stresses and the `PwBaseWorkChain` will check for this and return exit code 501.
+        2. The k-points mesh was defined as a density using `kpoints_distance` for the `base_relax` input name space,
+           and the unit cell has changed in such a way that regenerating the k-points mesh using this density increases
+           the mesh compared to the previous run.
 
-        If the maximum number of meta convergence iterations has been exceeded and convergence has not been reached, the
-        structure cannot be considered to be relaxed and the final scf should not be run.
+        If one of these two conditions is met, another relaxation should be run as long as the maximum number of meta
+        convergence iterations is not exceeded.
         """
-        return self.ctx.is_converged and 'final_scf_inputs' in self.ctx
+        # If no work chain has been run, we should at least run one
+        if 'base_relax_workchains' not in self.ctx:
+            return True
+
+        # If meta convergence is switched off we are done
+        if not self.ctx.meta_convergence:
+            self.report('Meta-convergence disabled: workchain completed after a single iteration.')
+            return False
+
+        # Stop if the maximum number of meta iterations has been reached
+        if self.ctx.iteration == self.inputs.max_meta_convergence_iterations.value:
+            self.report('Maximum number of meta convergence iterations reached.')
+            return False
+
+        base_relax_workchain = self.ctx.base_relax_workchains[-1]
+
+        # If the last work chain still found Pulay stresses in the final SCF, continue
+        pulay_exit_status = PwCalculation.exit_codes.ERROR_IONIC_CONVERGENCE_REACHED_EXCEPT_IN_FINAL_SCF.status
+        if base_relax_workchain.exit_status == pulay_exit_status:
+            self.report('Pulay stresses still present, running another geometry optimizatio.')
+            return True
+
+        # If the kpoints are defined as a density, make sure the kpoints mesh is the same for the new structur
+        if 'kpoints_distance' in self.ctx.relax_inputs:
+            input_kpts_mesh, _ = base_relax_workchain.base.links.get_outgoing(
+                link_label_filter='iteration_01'
+            ).first().node.inputs.kpoints.get_kpoints_mesh()
+            inputs_create_kpoints = {
+                'structure': base_relax_workchain.outputs.output_structure,
+                'distance': self.ctx.relax_inputs.kpoints_distance,
+                'force_parity': self.ctx.relax_inputs.get('kpoints_force_parity', orm.Bool(False)),
+                'metadata': {
+                    'store_provenance': False
+                }
+            }
+            new_kpts_mesh, _ = create_kpoints_from_distance(**inputs_create_kpoints).get_kpoints_mesh()
+            if tuple(input_kpts_mesh) != tuple(new_kpts_mesh):
```

Comment:
_EDIT: just saw your point in the PR description. Will leave this here because it contains some more thoughts on the exact criterion_.

Should we maybe here check only that the new grid is not denser? If the new required grid is _less_ dense, then there is no point in rerunning is there? The criterion for `kpoints_distance` that a _minimum_ density is guaranteed in order to guarantee a minimum precision.

One tricky thing I guess might be if you have non isometric meshes, such as `(2, 3, 1)` and `(3, 2, 1)`. Which one is denser? I guess there you would have to compute the actual distance between kpoints in 1/A for both both meshes in each direction and as soon as just one is larger for the new structure, we conclude we have to rerun.

Reply:
I'm not sure how common that case last case is, but it's a fair point. Maybe we can accept when the new mesh is denser along every direction. This would _not_ accept your last use case, but I think that's fine, and simpler than _really_ calculating the density.